### ISSUE TITLE
instagram website changed- "Non-retriable status code 403 while downloading"

### ISSUE THREAD

**Issue by Ectomind1990** [RipMeApp/ripme#484]
* Ripme version: 1.7.27
* Java version: 1.8
* Exact URL you were trying to rip when the problem occurred: (SFW) https://www.instagram.com/kutovakika/
* Please include any additional information about how to reproduce the problem:

It happens on other instagram accounts also. Its related to #384 which I raised in january

## Expected Behavior

Download photos and videos from all 537 posts on that account

## Actual Behavior

```
Downloading https://www.instagram.com/kutovakika/
Downloading https://scontent-lht6-1.cdninstagram.com/5B4467C7/t51.2885-15/sh0.08/e35/29093749_367590587052768_3182474189701382144_n.jpg
https://scontent-lht6-1.cdninstagram.com/5B4467C7/t51.2885-15/sh0.08/e35/29093749_367590587052768_3182474189701382144_n.jpg : Non-retriable status code 403 while downloading https://scontent-lht6-1.cdninstagram.com/5B4467C7/t51.2885-15/sh0.08/e35/29093749_367590587052768_3182474189701382144_n.jpg
Downloading https://scontent-lht6-1.cdninstagram.com/5B3456D8/t51.2885-15/sh0.08/e35/28754405_2041963605835948_1551946411904335872_n.jpg
https://scontent-lht6-1.cdninstagram.com/5B3456D8/t51.2885-15/sh0.08/e35/28754405_2041963605835948_1551946411904335872_n.jpg : Non-retriable status code 403 while downloading https://scontent-lht6-1.cdninstagram.com/5B3456D8/t51.2885-15/sh0.08/e35/28754405_2041963605835948_1551946411904335872_n.jpg
```

This is because [this is not an official URL](https://scontent-lht6-1.cdninstagram.com/5B4467C7/t51.2885-15/sh0.08/e35/29093749_367590587052768_3182474189701382144_n.jpg) coming from instagram. Ripme is trying to get the full resolution photo by removing sub strings of the official URL coming from instagram. But that trick doesn't always work any more- instagram have started changing it this week

**Comment by thinkpad4** [comment]
I noticed that too, was wondering what was going on

**Comment by cyian-1756** [comment]
> I noticed that too, was wondering what was going on

It looks like all URLs have to be signed now.

**Comment by thinkpad4** [comment]
@cyian-1756 I noticed that /t51.2885-15/e35/ appears to be consistent in the few different links I've tried

**Comment by cyian-1756** [comment]
I've managed to kinda fix this by downloading the display_url instead of getting the full sized image. It's not prefect (The image that is downloaded isn't the largest possible) but it until someone manages to crack how IG is generating the signature it's the best that can be done

**Comment by thinkpad4** [comment]
@cyian-1756 It appears it always be like this /5B719B26/t51.2885-15/e35/  the /5B719B26/ seems to be a random HEX number but the /t51.2885-15/e35/ is always the same, at least for now.  Can't you make it like /*/t51.2885-15/e35/  Replacing the HEX number with a * or something?

**Comment by cyian-1756** [comment]
> Can't you make it like /*/t51.2885-15/e35/ Replacing the HEX number with a * or something?

I can but that still causes a 404. As far as I can tell the vp/6900a5948df03361950a2164a4ba560d part of the url is a md5 hash of the URL + a salt. If I you change the url at all you need to generate a new hash, but I don't have the salt so I can't generate a new hash

**Comment by cyian-1756** [comment]
The image url for the image at https://www.instagram.com/p/BgrT-Gzh3F7/?taken-by=kutovakika is https://scontent-yyz1-1.cdninstagram.com/vp/6900a5948df03361950a2164a4ba560d/5B2E5099/t51.2885-15/s640x640/sh0.08/e35/28766860_558178931211208_8285564106707042304_n.jpg

Changing the image url to http doesn't cause the "Invalid URL signature" so it's safe to say that the "http" isn't part of the url that gets hashed

Making the sig longer or shorter returns a "Bad URL signature param" error so we can assume the sig is always 32 char long

**Comment by cyian-1756** [comment]
This is kinda fixed in 1.7.28 but I'm leaving this open encase someone figures out how to get full size images again

**Comment by thinkpad4** [comment]
@cyian-1756 Don't know if this would help but I was looking at the git of what appears to be the only working IG image downloader extension and they said "The right approach here seems to be fetching the image data and converting that to a blob which can be downloaded."  I don't know if you can do that in RipMe but I hope so

**Comment by cyian-1756** [comment]
> looking at the git of what appears to be the only working IG image downloader

Can you link me to that?

**Comment by thinkpad4** [comment]
@cyian-1756 https://github.com/ehmorris/Instagram-High-Resolution-Downloader

**Comment by thinkpad4** [comment]
@cyian-1756 Found another method that maybe you work with, it's from a user on Reddit.  https://www.reddit.com/r/Instagram/comments/86l6w8/instagram_profile_picture/dw9veal/

**Comment by Hrxn** [comment]
But this is for the profile picture only?

**Comment by thinkpad4** [comment]
The method on reddit gets the full sized HQ image

**Comment by cyian-1756** [comment]
@thinkpad4 

Am I misreading something or does the reddit method only work for profile images?

**Comment by thinkpad4** [comment]
OMG, I am SUCH a derp.  I didn't even see it saying Instagram Profile Pic in the title of that reddit post

**Comment by cyian-1756** [comment]
@thinkpad4 It's fine

**Comment by kevin51jiang** [comment]
Going back to the IG topic, would using the actual Facebook/Instagram API work?

**Comment by cyian-1756** [comment]
> Going back to the IG topic, would using the actual Facebook/Instagram API work?

Maybe, but AFAIK the facebook api is locked down right now because of that data scraping scandal from a while back and IG keeps changing their API.

> EDIT: Also thinking of embedding Selenium or something

We could and it's been done before, but it's a lot of extra weight on the end JAR for not that much benefit (Even most JS heavy sites can be scraped just using JSOUP)

**Comment by Ectomind1990** [comment]
> This is kinda fixed in 1.7.28 but I'm leaving this open encase someone figures out how to get full size images again

thanks for fixing and keeping us in the loop about what the deal is and thanks in general to the people making this tool

**Comment by Ectomind1990** [comment]
On gallery pages I'm still getting the full size images. Doing https://www.instagram.com/karencantuq/ 
Gets these files
2018_03_13_04_57_BgRVsdZgFB528764494_200825714018087_2503767454539317248_n.jpg
2018_03_13_04_57_BgRVsdZgFB528752003_195423554566125_7057308480591364096_n.jpg
They are both over 2mb

Does this help find a way to get full size from non gallery pages also?

**Comment by Hrxn** [comment]
What do you mean by non-gallery pages?

**Comment by cyian-1756** [comment]
This should be fixed in 1.7.36

**Comment by Ectomind1990** [comment]
A page with only 1 image like https://www.instagram.com/p/BgXj_xKA67E/ But I just tested it and 85.2kb is actually the full size image which is what ripme gets. I think? So actually this is fixed??

**Comment by Ectomind1990** [comment]
> This should be fixed in 1.7.36
thanks :+1:
